src/dd/Approximation.cpp

```diff
@@ -0,0 +1,246 @@
+      // If necessary, reset the lowest qubit number effected.
+      if (budget >= n.contribution) {
+        n.ptr->flags = FLAG_DELETE;
+        budget -= n.contribution;
+        min = std::min(min, n.ptr->v);
+        removeAbove(candidates, budget);
+        continue;
+      }
+
+      // Compute the contributions of the next layer.
+      for (std::size_t i = 0; i < RADIX; ++i) {
+        const vEdge& eRef = n.ptr->e[i];
+        const double contribution =
+            n.contribution * ComplexNumbers::mag2(eRef.w);
+
+        if (eRef.isTerminal()) {        // Don't add terminals to the queue.
+          if (budget >= contribution) { // They (potentially) can be deleted.
+            const uint16_t flag = (i == 0 ? FLAG_LEFT : FLAG_RIGHT);
+            candidates.emplace_front(n.ptr, flag, contribution);
+          }
+          continue;
+        }
+
+        c[eRef.p] += contribution;
+      }
+    }
+
+    Layer next{}; // Prioritize nodes for next iteration.
+    for (auto& [n, contribution] : c) {
+      next.emplace(n, contribution, budget);
+    }
+
+    if (next.empty()) { // Break early. Avoid std::move.
+      break;
+    }
+
+    curr = std::move(next);
+  }
+
+  // Lastly, check if any terminals can be deleted.
+  while (!candidates.empty()) {
+    const auto [n, flag, contribution] = candidates.front();
+    candidates.pop_front();
+    if (budget >= contribution) {
+      n->flags = FLAG_DELETE + flag;
+      budget -= contribution;
+      min = std::min(min, n->v);
+    }
+  }
+
+  if (meta != nullptr) {
+    rmeta.min = min;
+    rmeta.budgetLeft = budget;
+    *meta = rmeta;
+  }
+
+  // The final fidelity is the desired fidelity plus the unused budget.
+  return {fidelity + budget, min};
+}
+
+vEdge sweep(const vEdge& curr, const Qubit min, Lookup& l, Package& dd) {
+  vNode* n = curr.p;
+
+  // Nodes below v_{min} don't require rebuilding.
+  if (n->v < min) {
+    return curr;
+  }
+
+  // If a node is flagged, reset the flag and return a zero edge.
+  if (n->flags == FLAG_DELETE) {
+    n->flags = 0U;
+    return vEdge::zero();
+  }
+
+  // If a node has been visited once, return the already rebuilt node
+  // and set the edge weight accordingly.
+  if (l.find(n) != l.end()) {
+    vEdge eR = l.at(n);
+    eR.w = curr.w;
+    return eR;
+  }
+
+  // Otherwise traverse down to rebuild each non-terminal edge.
+  std::array<vEdge, RADIX> edges{};
+  for (std::size_t i = 0; i < RADIX; ++i) {
+    const vEdge& eRef = n->e[i];
+
+    if (eRef.isTerminal()) {
+      // Use zero edge for marked terminals.
+      const uint16_t flag = (i == 0 ? FLAG_LEFT : FLAG_RIGHT);
+      if (n->flags == FLAG_DELETE + flag) {
+        edges[i] = vEdge::zero();
+        continue;
+      }
+
+      edges[i] = eRef;
+      continue;
+    }
+
+    edges[i] = sweep(eRef, min, l, dd);
+  }
+
+  // Rebuild node and set its edge weight accordingly.
+  vEdge eR = dd.makeDDNode(n->v, edges);
+  eR.w = curr.w;
```

Comment:
This seems to work, but I am not quite sure I understand why at the moment.
Couldn't the makeDDNode function already return a non-one edge weight?
Why is the current edge weight involved here at all?
Probably just me sitting on the phone and overlooking something 😄